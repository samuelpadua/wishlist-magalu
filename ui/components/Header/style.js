import styled from 'styled-components'

export const Logo = styled.div`
  img {
    width: 100px;
  }
`