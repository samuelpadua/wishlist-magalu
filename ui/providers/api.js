import Axios from 'axios'

export const urls = {
  test: `http://localhost`,
  development: 'http://localhost:9000',
  production: 'http://localhost:9000'
}

const api = Axios.create({
  baseURL: urls[process.env.NODE_ENV],
  headers: {
    'Accept': 'application/json',
    'Content-Type': 'application/json'
  }
})

export default api
