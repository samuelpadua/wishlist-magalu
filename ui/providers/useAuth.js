import React, { createContext, useState, useContext, useEffect } from 'react'
import Cookies from 'js-cookie'
import Router from 'next/router'

import api from './api'

export const AuthContext = createContext({});

export function AuthProvider({ children }) {
  const [authenticated, setAuthenticated] = useState(null)
  const [loading, setLoading] = useState(true)
  const [customer, setCustomer] = useState(null)

  useEffect(() => {
    async function validateLoggedCustomer() {
      const token = Cookies.get('token')
      if (token) {
        api.defaults.headers.authorization = token

        const { data } = await api.get('/api/auth/me')

        if (data) {
          setCustomer(data)
          setAuthenticated(true)
        }
      }

      setLoading(false)
    }

    validateLoggedCustomer()
  }, [])

  async function login(email) {
    const response = await api.post('/api/auth', { email })
    const { data: { token } } = response

    if (token) {
      Cookies.set('token', token, { expires: 60 })
      api.defaults.headers.authorization = token

      const { data } = await api.get('/api/auth/me')

      setCustomer(data)
      setAuthenticated(true)
      setLoading(false)
    }
  }

  function logout() {
    Cookies.remove('token')
    setAuthenticated(false)
    setCustomer({})

    Router.push('/')
  }

  return (
    <AuthContext.Provider value={{ isAuthenticated: authenticated, login, customer, loading, logout }}>
      {children}
    </AuthContext.Provider>
  )
}

export function ProtectRoute(Component) {
  return () => {
    const { isAuthenticated, loading } = useAuth()

    useEffect(() => {
      if (!isAuthenticated && !loading) {
        Router.push('/entrar')
      }
    }, [loading, isAuthenticated])

    return (<Component {...arguments} />)
  }
}

export default function useAuth() {
  const context = useContext(AuthContext)

  return context
}

