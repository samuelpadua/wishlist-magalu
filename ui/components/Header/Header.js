import React, { Fragment } from 'react'
import { Flex, Box } from 'rebass'
import Link from 'next/link'
import { MdAccountCircle } from 'react-icons/md'
import { IoMdHeart } from 'react-icons/io'

import { Container } from '../Layout'
import { Logo } from './style'
import styled from 'styled-components'
import useAuth from '../../providers/useAuth'

const Nav = styled.nav`
  background: #1686FF;
`

const LoginLink = styled.div`
  a {
    padding-top: 4px;
    color: #ffffff !important;
    cursor: pointer;
  }
`

function Header() {
  const { isAuthenticated } = useAuth()

  return (
    <Nav>
      <Container>
        <Flex width={1}>
          <Box pt={22} pb="8px">
            <Link href="/">
              <Logo>
                <img src="https://lojas.magazineluiza.com.br/_nuxt/img/7064da1.png" />
              </Logo>
            </Link>
          </Box>
          <Flex justifyContent="flex-end" py={15} width={1}>
            {!isAuthenticated && (
              <LoginLink>
                <Link href="/entrar">
                  Entrar
                </Link>
              </LoginLink>
            )}
            {isAuthenticated && (
              <Fragment>
                <Box mr={10}>
                  <Link href="/favoritos" style={{ cursor: 'pointer' }}>
                    <IoMdHeart color="#ffffff" size="2em" />
                  </Link>
                </Box>
                <Box>
                  <Link href="/minha-conta" style={{ cursor: 'pointer' }}>
                    <MdAccountCircle color="#ffffff" size="2em" />
                  </Link>
                </Box>
              </Fragment>
            )}
          </Flex>
        </Flex>
      </Container>
    </Nav>
  )
}

export default Header
