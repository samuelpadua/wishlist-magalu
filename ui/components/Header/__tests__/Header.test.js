import React from 'react'
import renderer from 'react-test-renderer'
import 'jest-styled-components'
import { AuthProvider } from '../../../providers/useAuth'
import Header from '../Header'

describe('Header component', () => {
  it('should render no logged customer', () => {
    const tree = renderer
      .create(
        <AuthProvider>
          <Header />
        </AuthProvider>
      )
      .toJSON()

    expect(tree).toMatchSnapshot()
  })
})
