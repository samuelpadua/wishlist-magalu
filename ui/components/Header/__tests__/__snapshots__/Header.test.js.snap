// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header component should render no logged customer 1`] = `
.c1 {
  position: relative;
  margin: 0 auto !important;
  max-width: 1220px;
}

.c2 img {
  width: 100px;
}

.c0 {
  background: #1686FF;
}

.c3 a {
  padding-top: 4px;
  color: #ffffff !important;
  cursor: pointer;
}

<nav
  className="c0"
>
  <div
    className="c1 css-5g036t"
  >
    <div
      className="css-knfhfe"
    >
      <div
        className="css-1ainljp"
      >
        <div
          className="c2"
          onClick={[Function]}
          onMouseEnter={[Function]}
        >
          <img
            src="https://lojas.magazineluiza.com.br/_nuxt/img/7064da1.png"
          />
        </div>
      </div>
      <div
        className="css-165jgs1"
      >
        <div
          className="c3"
        >
          <a
            href="/entrar"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            Entrar
          </a>
        </div>
      </div>
    </div>
  </div>
</nav>
`;
